File: vane-velocity/src/main/java/org/oddlama/velocity/compat/event/VelocityCompatPendingConnection.java
package org.oddlama.velocity.compat.event;

import com.velocitypowered.api.proxy.InboundConnection;
import com.velocitypowered.api.proxy.Player;
import java.net.SocketAddress;
import java.util.Arrays;
import java.util.UUID;
import org.jetbrains.annotations.Nullable;
import org.oddlama.vane.proxycore.ProxyPendingConnection;
import org.oddlama.vane.proxycore.ProxyServer;

public class VelocityCompatPendingConnection implements ProxyPendingConnection {

    final InboundConnection connection;
    final String username;
    final UUID uuid;

    public VelocityCompatPendingConnection(InboundConnection connection, String username) {
        this.connection = connection;
        this.username = username;
        this.uuid = null;
    }

    public VelocityCompatPendingConnection(Player player) {
        this.connection = player;
        this.username = player.getUsername();
        this.uuid = player.getUniqueId();
    }

    @Override
    public String getName() {
        return username;
    }

    @Override
    public @Nullable UUID getUniqueId() {
        return uuid;
    }

    @Override
    public int getPort() {
        return connection.getVirtualHost().get().getPort();
    }

    @Override
    public SocketAddress getSocketAddress() {
        return connection.getVirtualHost().get();
    }

    @Override
    public boolean hasPermission(ProxyServer server, String... permission) {
        // Safe cast, we don't use this outside the case where this is constructed
        // with a Player
        return Arrays.stream(permission).anyMatch(perm -> ((Player) connection).hasPermission(perm));
    }
}

File: vane-velocity/src/main/kotlin/org/oddlama/velocity/compat/event/VelocityCompatPendingConnection.kt
package org.oddlama.velocity.compat.event

import com.velocitypowered.api.proxy.InboundConnection
import com.velocitypowered.api.proxy.Player
import org.oddlama.vane.proxycore.ProxyPendingConnection
import org.oddlama.vane.proxycore.ProxyServer
import java.net.SocketAddress
import java.util.*

class VelocityCompatPendingConnection : ProxyPendingConnection {
    val connection: InboundConnection
    override val name: String?
    override val uniqueId: UUID?

    constructor(connection: InboundConnection, username: String?) {
        this.connection = connection
        this.name = username
        this.uniqueId = null
    }

    constructor(player: Player) {
        this.connection = player
        this.name = player.username
        this.uniqueId = player.uniqueId
    }

    override val port: Int
        get() = connection.virtualHost.get().port

    override val socketAddress: SocketAddress
        get() = connection.virtualHost.get()

    override fun hasPermission(server: ProxyServer?, vararg permission: String?): Boolean {
        // Safe cast, we don't use this outside the case where this is constructed
        // with a Player
        return Arrays.stream<String?>(permission)
            .anyMatch { perm: String? -> (connection as Player).hasPermission(perm) }
    }
}
